bootstrap/kubeadm/internal/ignition/butane/butane.go

```diff
@@ -0,0 +1,414 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package butane generates bootstrap data in Ignition v3.4 format using Butane Transpiler.
+//
+// Butane configuration defined in this package will run kubeadm command by creating a /etc/kubeadm.sh script
+// file containing both pre and post kubeadm commands as well as the kubeadm command itself.
+//
+// /etc/kubeadm.sh script will be executed using kubeadm.service systemd unit, which will only happen
+// if /etc/kubeadm.yml file exists, which ensures the script will run only once, as by the end of the
+// script, /etc/kubeadm.yml is moved to /tmp filesystem, so it gets automatically cleaned up after a
+// reboot. This is to align the implementation with cloud-init, which places kubeadm configuration in
+// /tmp directory directly, which is not possible with Ignition.
+//
+// /etc/kubeadm.yml file contains generated kubeadm configuration and can be customized using pre kubeadm
+// commands if needed, as a replacement for Jinja templates supported by cloud-init, for example
+// using 'envsubst' or 'sed'.
+//
+// To override the behavior of kubeadm.service unit, one should create an override drop-in
+// using AdditionalConfig field. Data from this field takes precedence and will be merged with
+// configuration generated by the bootstrap provider, overriding already defined fields following the
+// merge strategy described in https://coreos.github.io/ignition/operator-notes/#config-merging.
+package butane
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/coreos/butane/config"
+	"github.com/coreos/butane/config/common"
+	ignition "github.com/coreos/ignition/v2/config/v3_4"
+	ignitionTypes "github.com/coreos/ignition/v2/config/v3_4/types"
+	"github.com/pkg/errors"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+)
+
+const (
+	butaneTemplate = `---
+variant: flatcar
```

Comment:
This is the variant & version of the static template only, the additionalConfig must specify it's own variant/version, both are translated to ignition v3.4.0 and the merge is done in ignition v3.4.0.

Reply:
That seem counter-intuitive, as `additionalConfig` name suggests it's _optional_ and so is the API semantics for this field. How do we _require_ it right now? Shouldn't we require this information from user and use it then for both base template and additional config?

Reply:
`additionalConfig` is still optional, the variant/version is required by butane when `additionalConfig` is not nil to transpile it to Ignition v3.4.0.

This version doesn't have to be the same as the one used in the static template as the merge is done in Ignition v3.4.0, so a user can add an  `additionalConfig`  in any variant/version they variant (as long as the couple can be transpiled to Ignition v3.4.0)

We could add a default variant/version if none is specified as it would work in most cases.

Reply:
I understand that, but if user specifies no `additionalConfig`, why should they be forced to use `flatcar` variant in this case? What if this variant is not compatible with their OS?